drivers/iio/proximity/vcnl3020.c

```diff
@@ -0,0 +1,338 @@
+		goto exit;
+	}
+
+	/* set low threshold value */
+	rc = i2c_smbus_write_byte_data(data->client, VCNL_PS_LO_THR_HI,
+				       (threshold >> 8) & 0xFF);
+	if (rc < 0) {
+		dev_err(dev, "Error set high threshold hreg %x", rc);
+		goto exit;
+	}
+
+	/* enable interrupt for exceeding low/high thresholds */
+	rc = i2c_smbus_write_byte_data(data->client, VCNL_PS_ICR,
+				       (count_exceed << ICR_COUNT_EXCEED) |
+				       ICR_THRES_EN);
+	if (rc < 0) {
+		dev_err(dev, "Error set interrupt control reg %x", rc);
+		goto exit;
+	}
+
+	rc = i2c_smbus_write_byte_data(data->client, VCNL_COMMAND,
+				       (VCNL_PS_EN | VCNL_PS_SELFTIMED_EN) &
+				       (VCNL_PS_OD ^ 0xff));
+	if (rc < 0) {
+		dev_err(dev, "Error set interrupt control reg %x", rc);
+		goto exit;
+	}
+
+	rc = i2c_smbus_read_byte_data(data->client, VCNL_COMMAND);
+	dev_err(dev, "command data %x", rc);
+
+	return 0;
+exit:
+	return rc;
+};
+
+#ifdef CONFIG_SENSORS_VCNL3020
+
+bool vcnl3020_intrusion(struct vcnl3020_data *data)
+{
+	s32 isr;
+
+	mutex_lock(&data->vcnl3020_lock);
+	isr = i2c_smbus_read_byte_data(data->client, VCNL_ISR);
+	if (isr < 0) {
+		dev_err(&data->client->dev, "Error reading interrupt status %x",
+			isr);
+	}
+	mutex_unlock(&data->vcnl3020_lock);
+
+	return !!(isr & BIT(INT_TH_LOW));
+}
+EXPORT_SYMBOL_GPL(vcnl3020_read_isr);
+
+int32_t vcnl3020_clear_interrupts(struct vcnl3020_data *data)
+{
+	s32 rc;
+
+	mutex_lock(&data->vcnl3020_lock);
+	rc = i2c_smbus_write_byte_data(data->client, VCNL_ISR,
+				       INT_TH_HI | INT_TH_LOW | INT_PROX_READY);
+	mutex_unlock(&data->vcnl3020_lock);
+
+	return rc;
+}
+EXPORT_SYMBOL_GPL(vcnl3020_clear_interrupts);
+
+#endif
+
+static int32_t vcnl3020_measure_proximity(struct vcnl3020_data *data,
+					  int32_t *val)
+{
+	s32 tries = 20;
+	s32 rc = 0;
+	s32 cmdreg = 0;
+
+	mutex_lock(&data->vcnl3020_lock);
+	/* store command register state before measurement */
+	cmdreg = i2c_smbus_read_byte_data(data->client, VCNL_COMMAND);
+	if (cmdreg < 0)
+		goto fail;
+
+	rc = i2c_smbus_write_byte_data(data->client, VCNL_COMMAND,
+				       VCNL_PS_OD);
+	if (rc < 0)
+		goto fail;
+
+	/* wait for data to become ready */
+	while (tries--) {
+		rc = i2c_smbus_read_byte_data(data->client, VCNL_COMMAND);
+		if (rc < 0)
+			goto fail;
+		if (rc & VCNL_PS_RDY)
+			break;
+		msleep(20); /* measurement takes up to 100 ms */
+	}
+
+	if (tries < 0) {
+		dev_err(&data->client->dev,
+			"vcnl3020_measure() failed, data is not ready");
+		rc = -EIO;
+		goto fail;
+	}
+
+	rc = i2c_smbus_read_byte_data(data->client, VCNL_PS_RESULT_HI);
+	if (rc < 0)
+		goto fail;
+	*val = (rc & 0xff) << 8;
+	dev_dbg(&data->client->dev, "result high byte 0x%x", rc);
+
+	rc = i2c_smbus_read_byte_data(data->client, VCNL_PS_RESULT_LO);
+	if (rc < 0)
+		goto fail;
+	*val |= rc & 0xff;
+	dev_dbg(&data->client->dev, "result low byte 0x%x", rc);
+
+	rc = i2c_smbus_write_byte_data(data->client, VCNL_COMMAND, cmdreg);
+	if (rc < 0)
+		goto fail;
+
```

Comment:
Refracting is needed. You can return rc an any way.

Reply:
I didn't quite understand what you meant here, but if you were saying that Ivan needed to put a `return rc` right here, then you're wrong, it would be a violation of [rule 7 of Linux Kernel Coding Style](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#centralized-exiting-of-functions). Centralized exiting of functions is preffered by Linus (and me).

Reply:
I mean replace
```
	mutex_unlock(&data->vcnl3020_lock);

	return 0;
fail:
	mutex_unlock(&data->vcnl3020_lock);
	return rc;
```
with
```
fail:
	mutex_unlock(&data->vcnl3020_lock);
	return rc;
```
or something like that